File: spark/src/main/scala/geotrellis/spark/io/index/KeyIndexMethod.scala
package geotrellis.spark.io.index

import geotrellis.spark._

trait KeyIndexMethod[K] extends Serializable {
  /** Helper method to get the resolution of a dimension. Takes the ceiling. */
  def resolution(max: Double, min: Double): Int = {
    val length = max - min + 1
    math.ceil(scala.math.log(length) / scala.math.log(2)).toInt
  }

  def createIndex(keyBounds: KeyBounds[K]): KeyIndex[K]
}

File: geotrellis-python/geotrellis/spark/io/index/KeyIndexMethod.py
import math

class _KeyIndexMethodMeta(object):
    items = {}
    def __getitem__(self, key_type):
        if key_type in self.items.keys():
            return self.items[key_type]
        class tempo(_KeyIndexMethod):
            _key_type = key_type
        self.items[key_type] = tempo
        return tempo

KeyIndexMethod = _KeyIndexMethodMeta()

class _KeyIndexMethod(object):

    _key_type = None

    def resolution(self, _max, _min):
        length = _max - _min + 1
        return int(math.ceil(math.log(length) / math.log(2)))

    def createIndex(self, keyBounds):
        pass
